### ISSUE TITLE
Missing socket.session once upgraded to 0.4.4

### ISSUE THREAD

**Issue by twhtanghk** [fullstackers/socket.io-events#11]
Once upgrade to 0.4.4, the function below return the socket with emit method only and no more session available. Any suggestion or hints.

```
function Socket (sock) {
  if (!(this instanceof Socket)) return new Socket(sock);
  this.__proto__ = sock;

  this.emit = function() {
    debug('debug wrapped');
    return sock.emit.apply(sock, slice.call(arguments));
  };
}
```

**Comment by NathanGRomano** [comment]
I would be glad to troubleshoot and get this resolved.  Do you have a sample piece of code I can test with?

**Comment by NathanGRomano** [comment]
@smart--petea The change you made may have had an impact on @twhtanghk code.  Lets see if we can't figure this one out, we may have to revert.  Or maybe make this change

``` javascript
var emit = sock.emit;
this.emit = function () {
  debug('debug wrapped');
  return emit.apply(sock, slice.call(arguments));
}
```

@twhtanghk maybe you could try this out and see if it resolves the issue?

**Comment by smart--petea** [comment]
I made that change. But, what is the reason? 
`sock` reference dissapear and we must preserve the `emit` reference?

**Comment by NathanGRomano** [comment]
The point of wrapping the socket.io `socket` is that multiple concurrent handlers maybe be interacting with it.  We wrap it up into our own `Socket` class which is responsible for delegating the `emit()` call to the internal `socket`.  The wrapper method is used internally by the router in order to stop propagation of the routing whenever the `emit()` event is called.  I was wrong in my statement above, however, I think your pull request is good.

**Comment by NathanGRomano** [comment]
@twhtanghk let me know if @smart--petea fix has resolved the issue, thanks!

**Comment by twhtanghk** [comment]
Not yet resolved. When I put the statement "this.sock = sock;" back, everything ok.

```
function Socket (sock) {
  if (!(this instanceof Socket)) return new Socket(sock);
  this.__proto__ = sock;
  this.sock = sock;

  var self = this;
  this.emit = function() {
    debug('debug wrapped');
    return sock.emit.apply(self, slice.call(arguments));
  };
}

```

**Comment by smart--petea** [comment]
There are not in route.js the code that use the `sock` attribute of socket. Maybe @twhtanghk  access the true socket in his own code by `sock.sock`. In this case it's enough to use only `sock`, it inherit with `__proto__` help all properties of true socket.

**Comment by NathanGRomano** [comment]
@twhtanghk by just doing `sock` do you get the same results as you used to by doing `sock.sock`?

**Comment by yads** [comment]
We need the wrapped socket in sock back as well. What was the reason for getting rid of it?

**Comment by NathanGRomano** [comment]
Yea we need the wrapped sock back it was critical

**Comment by NathanGRomano** [comment]
I will get code reverted back to the stable version.  @smart--petea There is a reason the socket was wrapped.  The socket instance is shared amongst multiple possible events.  IF we manipulate the **real** socket instance and not the wrapper, it could have side effects.  These side effects I believe are what our users are experience.

**Comment by twhtanghk** [comment]
In my case, only sock.emit is available and no more other method or properties available. That's why I need to put the statement back "this.sock = sock;".

**Comment by smart--petea** [comment]
ok.

**Comment by NathanGRomano** [comment]
@twhtanghk Are you still using version v0.4.2?

**Comment by twhtanghk** [comment]
Yes, with version exact match to 0.4.2.

**Comment by NathanGRomano** [comment]
We will get the fix in and deploy to version 0.4.6 today

**Comment by NathanGRomano** [comment]
@twhtanghk @smart--petea 

Version 0.4.6 has been deployed.  Let me know if there are any issues.
